lib/upgrade/etcd/conversionv1v3/rule.go

```diff
@@ -0,0 +1,266 @@
+func ruleBackendToAPI(br model.Rule) apiv3.Rule {
+	var icmp, notICMP *apiv3.ICMPFields
+	if br.ICMPCode != nil || br.ICMPType != nil {
+		icmp = &apiv3.ICMPFields{
+			Code: br.ICMPCode,
+			Type: br.ICMPType,
+		}
+	}
+	if br.NotICMPCode != nil || br.NotICMPType != nil {
+		notICMP = &apiv3.ICMPFields{
+			Code: br.NotICMPCode,
+			Type: br.NotICMPType,
+		}
+	}
+
+	srcNets := br.AllSrcNets()
+	var srcNetsStr []string
+	for _, net := range srcNets {
+		srcNetsStr = append(srcNetsStr, net.String())
+	}
+
+	dstNets := br.AllDstNets()
+	var dstNetsStr []string
+	for _, net := range dstNets {
+		dstNetsStr = append(dstNetsStr, net.String())
+	}
+
+	notSrcNets := br.AllNotSrcNets()
+	var notSrcNetsStr []string
+	for _, net := range notSrcNets {
+		notSrcNetsStr = append(notSrcNetsStr, net.String())
+	}
+
+	notDstNets := br.AllNotDstNets()
+	var notDstNetsStr []string
+	for _, net := range notDstNets {
+		notDstNetsStr = append(notDstNetsStr, net.String())
+	}
+
+	// Tags are deprecated in v3.0+, so we convert Tags to selectors.
+	// For example:
+	// A rule that looks like this with Calico v1 API:
+	// apiv1.Rule{
+	//	Action: "allow",
+	//	Source: apiv1.EntityRule{
+	//		Tag:      "tag1",
+	//		Selector: "label1 == 'value1' || make == 'cake'",
+	//	},
+	//}
+	// That rule will be converted to the following for Calico v3 API:
+	// apiv3.Rule{
+	//	Action: "allow",
+	//	Source: apiv3.EntityRule{
+	//		Selector: "(label1 == 'value1' || make == 'cake') && tag1 == ''",
+	//	},
+	//}
+	srcSelector := br.SrcSelector
+	if br.SrcTag != "" {
+		srcSelector = fmt.Sprintf("(%s) && %s == ''", br.SrcSelector, br.SrcTag)
+	}
+
+	dstSelector := br.DstSelector
+	if br.DstTag != "" {
+		dstSelector = fmt.Sprintf("(%s) && %s == ''", br.DstSelector, br.DstTag)
+	}
+
+	notSrcSelector := br.NotSrcSelector
+	if br.NotSrcTag != "" {
+		notSrcSelector = fmt.Sprintf("(%s) && %s == ''", br.NotSrcSelector, br.NotSrcTag)
+	}
+
+	notDstSelector := br.NotDstSelector
+	if br.NotDstTag != "" {
+		notDstSelector = fmt.Sprintf("(%s) && %s == ''", br.NotDstSelector, br.NotDstTag)
+	}
+
+	v3Protocol := numorstring.ProtocolV3FromProtocolV1(*br.Protocol)
+
+	return apiv3.Rule{
+		Action:      ruleActionToV3API(br.Action),
+		IPVersion:   br.IPVersion,
+		Protocol:    &v3Protocol,
+		ICMP:        icmp,
+		NotProtocol: br.NotProtocol,
+		NotICMP:     notICMP,
+		Source: apiv3.EntityRule{
+			Nets:        srcNetsStr,
+			Selector:    srcSelector,
+			Ports:       br.SrcPorts,
+			NotNets:     notSrcNetsStr,
+			NotSelector: notSrcSelector,
+			NotPorts:    br.NotSrcPorts,
+		},
+
+		Destination: apiv3.EntityRule{
+			Nets:        dstNetsStr,
+			Selector:    dstSelector,
+			Ports:       br.DstPorts,
+			NotNets:     notDstNetsStr,
+			NotSelector: notDstSelector,
+			NotPorts:    br.NotDstPorts,
+		},
+	}
+}
+
+// ruleActionAPIToBackend converts the rule action field value from the API
+// value to the equivalent backend value.
+func ruleActionAPIToBackend(action string) string {
```

Comment:
I think you need to add in version info to all of these - is this v1API to v1Backend?